File: src/apps/mobile/shared/core-feature/src/commonMain/kotlin/com/openbitfun/mobile/core/feature/session/SessionRecordReplica.kt
package com.openbitfun.mobile.core.feature.session

import com.openbitfun.mobile.core.domain.ChatMessage
import com.openbitfun.mobile.core.protocol.*
import kotlinx.serialization.json.*

/** Revisioned canonical records retain the complete host DTO; rendering is a mobile adapter. */
internal class SessionRecordReplica(private val sessionId: String) {
    private data class Versioned(val revision: Long, val value: JsonObject, val authoritative: Boolean = false)
    private val turns = linkedMapOf<String, Versioned>()
    private val rounds = linkedMapOf<String, Versioned>()
    private val items = linkedMapOf<String, Versioned>()
    private val controls = mutableMapOf<String, Pair<String, RemoteToolStatusResponse>>()

    /**
     * True until the stream has delivered a record of its own.
     *
     * An empty replica has nothing to say about the transcript, and rendering it
     * would erase a timeline the cache or the history fetch already filled.
     */
    val isEmpty: Boolean get() = turns.isEmpty()
    fun applyControl(payload: JsonObject) {
        val turn = payload.string("turnId")
        val event = payload["toolEvent"] as? JsonObject ?: return
        val id = event.string("tool_id")
        if (turn.isEmpty() || id.isEmpty()) return
        when (event.string("event_type")) {
            "ConfirmationNeeded" -> {
                controls[id] = turn to RemoteToolStatusResponse(id = id, name = event.string("tool_name"), status = "pending_confirmation", toolInput = event["params"])
                touchTurn(turn)
            }
            "Confirmed", "Completed", "Failed", "Cancelled", "Rejected" -> {
                val owner = controls.remove(id)?.first
                if (owner != null) touchTurn(owner)
            }
        }
    }
    private val recordVersions = mutableMapOf<String, Long>()
    private val tombstones = mutableMapOf<String, Long>()
    private val itemRounds = mutableMapOf<String, String>()

    /**
     * Messages already rendered for a turn, kept until that turn changes again.
     *
     * Every record restates the whole session and the timeline reads all of it, so
     * rendering used to re-parse every message of every turn — the entire loaded
     * transcript, contents and all — for each record that arrived, on the thread
     * that draws the screen. A turn whose records and controls have not changed
     * renders to the same pair of messages, and handing back the same instances
     * also lets everything downstream compare them by identity instead of by
     * content.
     *
     * [turnGenerations] is the whole correctness argument: it is bumped by every
     * mutation that can change what a turn renders to, so a matching generation
     * means the cached pair is still the current one.
     */
    private data class RenderedTurn(val generation: Long?, val messages: List<ChatMessage>)

    private val renderedTurns = mutableMapOf<String, RenderedTurn>()
    private val turnGenerations = mutableMapOf<String, Long>()
    private var turnGenerationSeq = 0L

    /** Records the next render of [turnId] as different from the cached one. */
    private fun touchTurn(turnId: String) {
        turnGenerations[turnId] = ++turnGenerationSeq
    }

    private fun put(map: MutableMap<String, Versioned>, id: String, revision: Long, value: JsonObject) {
        if (revision > (map[id]?.revision ?: -1)) map[id] = Versioned(revision, value)
    }
    /** A round or item record repeats its parent turn so a partial page still reads,
     * but that copy is a header, not the turn's own state: the turn record is what the
     * user message and its attachment pixels come from, so a header fills a turn it has
     * never seen and never overwrites one. */
    private fun putTurn(id: String, revision: Long, value: JsonObject, authoritative: Boolean) {
        val stored = turns[id]
        if (stored != null && (revision <= stored.revision || (!authoritative && stored.authoritative))) return
        turns[id] = Versioned(revision, value, authoritative)
    }
    fun apply(payload: JsonObject) {
        check(payload.string("sessionId") == sessionId) { "Session record binding mismatch" }
        val revision = payload.getValue("revision").jsonPrimitive.long
        check(revision in 1..9007199254740991L) { "Invalid session record revision" }
        val recordId = payload.string("id")
        check(recordId.substringBefore('/') in setOf("turn", "round", "item") && recordId.substringAfter('/', "").isNotEmpty()) { "Invalid session record identity" }
        if (revision <= (recordVersions[recordId] ?: -1)) return
        if (payload["deleted"]?.jsonPrimitive?.booleanOrNull == true) {
            recordVersions[recordId] = revision
            if (revision > (tombstones[recordId] ?: -1)) tombstones[recordId] = revision
            // A deletion retires a turn, one of its rounds, or one of its items.
            // Only a whole turn is named by its own record id, so anything else
            // falls back to rendering every turn again.
            val named = recordId.removePrefix("turn/").takeIf { recordId.startsWith("turn/") }
            val owner = named ?: (payload["turn"] as? JsonObject)?.string("turnId")
            if (owner.isNullOrEmpty()) renderedTurns.clear() else touchTurn(owner)
            return
        }
        if (revision <= (tombstones[recordId] ?: -1)) return
        val turn = payload.getValue("turn").jsonObject
        check(turn.string("sessionId") == sessionId) { "Turn binding mismatch" }
        val turnId = turn.string("turnId")
        check(turnId.isNotEmpty()) { "Missing turn identity" }
        val recordRound = payload["round"] as? JsonObject
        val recordItem = payload["item"] as? JsonObject
        check(recordItem == null || recordRound != null) { "Item has no round" }
        val expectedId = if (recordItem != null) "item/" + recordItem.getValue("data").jsonObject.string("id") else if (recordRound != null) "round/" + recordRound.string("id") else "turn/$turnId"
        check(recordId == expectedId) { "Session record identity mismatch" }
        val roundId = recordRound?.string("id")
        if (recordRound != null) {
            check(recordRound.string("turnId") == turnId) { "Round binding mismatch" }
        }
        val itemId = recordItem?.getValue("data")?.jsonObject?.string("id")
        if (itemId != null) {
            check(itemRounds[itemId] == null || itemRounds[itemId] == roundId) { "Item parent changed" }
        }

        // Validate the complete ancestry before changing either data or replay fences.
        // A rejected record must leave a corrected delivery at the same revision usable.
        recordVersions[recordId] = revision
        putTurn(turnId, revision, turn, recordRound == null)
        if (turns[turnId]?.value?.string("status") != "inprogress") controls.entries.removeAll { it.value.first == turnId }
        if (recordRound != null && roundId != null) {
            put(rounds, roundId, revision, recordRound)
            if (recordItem != null && itemId != null) {
                itemRounds[itemId] = roundId
                put(items, itemId, revision, recordItem)
            }
        }
        touchTurn(turnId)
    }

    /**
     * Groups every round and item under its parent once per rebuild.
     *
     * Reading the transcript used to filter the whole round map for each turn
     * and the whole item map for each round, so a session paid for its own
     * length twice over on every record that arrived. Long sessions spend that
     * cost on the thread that draws them, which is where the scrolling went.
     */
    private data class Children(
        val roundsByTurn: Map<String, List<Pair<String, Versioned>>>,
        val itemsByRound: Map<String, List<Pair<String, Versioned>>>,
    )

    private fun children(): Children {
        val roundsByTurn = mutableMapOf<String, MutableList<Pair<String, Versioned>>>()
        rounds.forEach { (id, record) ->
            roundsByTurn.getOrPut(record.value.string("turnId")) { mutableListOf() }.add(id to record)
        }
        val itemsByRound = mutableMapOf<String, MutableList<Pair<String, Versioned>>>()
        items.forEach { (id, record) ->
            val roundId = itemRounds[id] ?: return@forEach
            itemsByRound.getOrPut(roundId) { mutableListOf() }.add(id to record)
        }
        return Children(roundsByTurn, itemsByRound)
    }

    fun messages(): List<ChatMessage> {
        val index = children()
        val visible = turns.filter { (id, record) -> record.revision > (tombstones["turn/$id"] ?: -1) }
            .values.sortedBy { it.value.number("turnIndex") }
        val messages = ArrayList<ChatMessage>(visible.size * 2)
        visible.forEach { record ->
            val turnId = record.value.string("turnId")
            val cached = renderedTurns[turnId]
            val generation = turnGenerations[turnId]
            if (cached != null && cached.generation == generation) {
                messages += cached.messages
            } else {
                val rendered = renderTurn(record.value, turnId, index)
                // renderTurn may retire stale control entries and advance the
                // generation; cache against the post-render value.
                renderedTurns[turnId] = RenderedTurn(turnGenerations[turnId], rendered)
                messages += rendered
            }
        }
        if (renderedTurns.size > visible.size) renderedTurns.keys.retainAll(visible.mapTo(mutableSetOf()) { it.value.string("turnId") })
        return messages
    }

    /** Renders one turn, the expensive half of [messages]. */
    private fun renderTurn(turn: JsonObject, turnId: String, index: Children): List<ChatMessage> {
        val user = turn.getValue("userMessage").jsonObject
        val turnFence = tombstones["turn/$turnId"] ?: -1
        val children = index.roundsByTurn[turnId].orEmpty().filter { (id, record) -> record.revision > maxOf(turnFence, tombstones["round/$id"] ?: -1) }
            .sortedBy { it.second.value.number("roundIndex") }.flatMap { (currentRoundId, _) ->
                val roundFence = maxOf(turnFence, tombstones["round/$currentRoundId"] ?: -1)
                index.itemsByRound[currentRoundId].orEmpty().filter { (id, record) -> record.revision > maxOf(roundFence, tombstones["item/$id"] ?: -1) }.map { it.second.value }
                    .filter { it.getValue("data").jsonObject.string("status") !in setOf("superseded", "retry_superseded") }
                    .sortedWith(compareBy({ it.getValue("data").jsonObject.number("orderIndex") }, { it.getValue("data").jsonObject.number("timestamp") }))
            }
        val rendered = children.map { item ->
            val data = item.getValue("data").jsonObject
            val type = item.string("type")
            val result = data["toolResult"] as? JsonObject
            ChatMessageItemResponse(type = type, content = data.string("content"), isSubagent = data["isSubagentItem"]?.jsonPrimitive?.booleanOrNull == true || data.string("subagentSessionId").isNotEmpty(),
                tool = if (type != "tool") null else RemoteToolStatusResponse(
                    id = (data["toolCall"] as? JsonObject)?.string("id")?.takeIf { it.isNotEmpty() } ?: data.string("id"), name = data.string("toolName"),
                    status = data.string("status").takeIf { it.isNotEmpty() }
                        ?: if (result == null) "running" else if (result["success"]?.jsonPrimitive?.booleanOrNull == true) "completed" else "failed",
                    toolInput = (data["toolCall"] as? JsonObject)?.get("input"), toolOutput = result?.get("result"),
                    errorPreview = result?.string("error"),
                    startMs = data["startTime"]?.jsonPrimitive?.longOrNull,
                    durationMs = data["durationMs"]?.jsonPrimitive?.longOrNull
                        ?: result?.get("durationMs")?.jsonPrimitive?.longOrNull))
        }
        val superseded = rendered.mapNotNull { it.tool }.filter { it.status in setOf("completed", "failed", "cancelled", "rejected", "skipped") }
        if (superseded.isNotEmpty()) {
            var controlsChanged = false
            superseded.forEach { if (controls.remove(it.id) != null) controlsChanged = true }
            // The cached render of a turn is only valid until its controls change.
            if (controlsChanged) touchTurn(turnId)
        }
        val controlTools = controls.values.filter { it.first == turnId }.map { it.second }
        val shownItems = rendered.map { item -> item.tool?.id?.let { id -> controlTools.firstOrNull { it.id == id } }?.let { item.copy(tool = it) } ?: item } +
            controlTools.filter { tool -> rendered.none { it.tool?.id == tool.id } }.map { ChatMessageItemResponse(type = "tool", tool = it) }
        return listOf(RemoteResponseMapper.chatMessage(ChatMessageResponse(id = user.string("id"), role = "user", content = user.string("content"), turnId = turnId, metadata = user["metadata"], images = userImages(user), timestamp = user.string("timestamp"))),
            RemoteResponseMapper.chatMessage(ChatMessageResponse(id = "${turnId}_assistant", role = "assistant", turnId = turnId,
                content = rendered.filter { it.type == "text" && it.isSubagent != true }.joinToString("") { it.content.orEmpty() },
                thinking = rendered.filter { it.type == "thinking" && it.isSubagent != true }.joinToString("") { it.content.orEmpty() },
                items = shownItems, status = when (turn.string("status")) { "inprogress" -> "streaming"; "error" -> "failed"; else -> turn.string("status") }, error = turn.string("error"), metadata = turn)))
    }

    /** Attachments are recorded with the turn; one that kept only a host path has no
     * pixels to hand a client that cannot reach that filesystem. */
    private fun userImages(user: JsonObject): List<ImageAttachment> =
        (((user["metadata"] as? JsonObject)?.get("images") as? JsonArray) ?: emptyList()).mapNotNull { entry ->
            val image = entry as? JsonObject ?: return@mapNotNull null
            image.string("data_url").takeIf { it.isNotEmpty() }?.let { ImageAttachment(name = image.string("name"), dataUrl = it) }
        }
    private fun JsonObject.string(key: String): String = (get(key) as? JsonPrimitive)?.contentOrNull.orEmpty()
    private fun JsonObject.number(key: String): Long = (get(key) as? JsonPrimitive)?.longOrNull ?: 0
}

File: src/shared/relay-transport/SessionRecordReplica.ts
/** Stable structural record contract; extra runtime fields survive spread/materialization. */
export interface TextItemData { id: string; orderIndex?: number; }
export interface ThinkingItemData { id: string; orderIndex?: number; }
export interface ToolItemData { id: string; orderIndex?: number; }
export interface ModelRoundData { id: string; turnId: string; roundIndex: number; textItems: TextItemData[]; thinkingItems?: ThinkingItemData[]; toolItems: ToolItemData[]; }
export interface DialogTurnData { turnId: string; sessionId: string; modelRounds: ModelRoundData[]; }

export interface SessionRecordUpsert {
  sessionId: string;
  id: string;
  revision: number;
  deleted?: false;
  turn: Omit<DialogTurnData, 'modelRounds'>;
  round?: Omit<ModelRoundData, 'textItems' | 'thinkingItems' | 'toolItems'>;
  item?: { type: 'text'; data: TextItemData }
    | { type: 'thinking'; data: ThinkingItemData }
    | { type: 'tool'; data: ToolItemData };
}
export type SessionRecord = SessionRecordUpsert | { sessionId: string; id: string; revision: number; deleted: true };
export interface SessionRecordChange { turnId: string; turn: DialogTurnData | null; }

type Versioned<T> = { revision: number; value: T };
type RoundEntry = {
  header: Versioned<NonNullable<SessionRecordUpsert['round']>>;
  items: Map<string, Versioned<NonNullable<SessionRecordUpsert['item']>>>;
};
type TurnEntry = {
  header: Versioned<SessionRecordUpsert['turn']>;
  rounds: Map<string, RoundEntry>;
};

/**
 * Live and fetched records enter the same reducer, as in Happy's message map.
 * Runtime revisions also make newest-first history safe for mutable blocks.
 * This instance belongs to one (device surface, session), never a global ID.
 */
export class SessionRecordReplica {
  private readonly records = new Map<string, number>();
  private readonly turns = new Map<string, TurnEntry>();
  private readonly deleted = new Map<string, number>();

  constructor(readonly sessionId: string) {}

  apply(record: SessionRecord): SessionRecordChange | null {
    if (record.sessionId !== this.sessionId || (!record.deleted && record.turn.sessionId !== this.sessionId)) {
      throw new Error('Session record belongs to another session');
    }
    if (!Number.isSafeInteger(record.revision) || record.revision < 1) {
      throw new Error('Invalid session record revision');
    }
    if ((this.records.get(record.id) ?? 0) >= record.revision) return null;
    if (record.deleted) {
      if (!/^(turn|round|item)\/.+/.test(record.id)) throw new Error('Invalid session record identity');
      this.records.set(record.id, record.revision);
      this.deleted.set(record.id, record.revision);
      for (const [turnId, turn] of this.turns) {
        if (record.id === `turn/${turnId}` || [...turn.rounds.entries()].some(([roundId, round]) => (
          record.id === `round/${roundId}` || [...round.items.keys()].some(itemId => record.id === `item/${itemId}`)
        ))) return this.change(turnId, turn);
      }
      return record.id.startsWith('turn/') ? { turnId: record.id.slice(5), turn: null } : null;
    }
    const turnId = record.turn.turnId;
    if (!turnId || (record.round && record.round.turnId !== turnId) || (record.item && !record.round)) {
      throw new Error('Invalid session record ancestry');
    }
    const expectedId = record.item ? `item/${record.item.data.id}`
      : record.round ? `round/${record.round.id}` : `turn/${turnId}`;
    if (record.id !== expectedId) throw new Error('Invalid session record identity');

    let turn = this.turns.get(turnId);
    if (!turn) {
      turn = { header: { revision: record.revision, value: record.turn }, rounds: new Map() };
      this.turns.set(turnId, turn);
    } else if (record.revision > turn.header.revision) {
      turn.header = { revision: record.revision, value: record.turn };
    }
    if (record.round) {
      let round = turn.rounds.get(record.round.id);
      if (!round) {
        round = { header: { revision: record.revision, value: record.round }, items: new Map() };
        turn.rounds.set(record.round.id, round);
      } else if (record.revision > round.header.revision) {
        round.header = { revision: record.revision, value: record.round };
      }
      if (record.item) {
        const existing = round.items.get(record.item.data.id);
        if (!existing || existing.revision < record.revision) {
          round.items.set(record.item.data.id, { revision: record.revision, value: record.item });
        }
      }
    }
    this.records.set(record.id, record.revision);
    return this.change(turnId, turn);
  }

  private change(turnId: string, turn: TurnEntry): SessionRecordChange {
    return { turnId, turn: turn.header.revision <= (this.deleted.get(`turn/${turnId}`) ?? 0)
      ? null : this.materialize(turnId, turn) };
  }

  private materialize(turnId: string, turn: TurnEntry): DialogTurnData {
    const turnFence = this.deleted.get(`turn/${turnId}`) ?? 0;
    return {
      ...turn.header.value,
      modelRounds: [...turn.rounds.values()].filter(round => round.header.revision > Math.max(turnFence, this.deleted.get(`round/${round.header.value.id}`) ?? 0)).map(round => {
        const roundFence = Math.max(turnFence, this.deleted.get(`round/${round.header.value.id}`) ?? 0);
        const result: ModelRoundData = { ...round.header.value, textItems: [], thinkingItems: [], toolItems: [] };
        for (const { value: item, revision } of round.items.values()) {
          if (revision <= Math.max(roundFence, this.deleted.get(`item/${item.data.id}`) ?? 0)) continue;
          if (item.type === 'text') result.textItems.push(item.data);
          else if (item.type === 'thinking') result.thinkingItems!.push(item.data);
          else result.toolItems.push(item.data);
        }
        const order = (a: { orderIndex?: number }, b: { orderIndex?: number }): number => (a.orderIndex ?? 0) - (b.orderIndex ?? 0);
        result.textItems.sort(order);
        result.thinkingItems!.sort(order);
        result.toolItems.sort(order);
        return result;
      }).sort((a, b) => a.roundIndex - b.roundIndex),
    };
  }
}
